semanticdb-javac/src/main/java/com/sourcegraph/semanticdb_javac/SemanticdbTaskListener.java

```diff
@@ -91,13 +112,60 @@ public static Path absolutePathFromUri(SemanticdbJavacOptions options, URI uri)
       } else {
         throw new IllegalArgumentException("unsupported URI: " + uri);
       }
+    } else if (options.uriScheme == UriScheme.BAZEL) {
+      String toString = file.toString();
+      // This solution is hacky, and it would be very nice to use a dedicated API instead.
+      // The Bazel Java compiler constructs `SimpleFileObject` with a "user-friendly" name that
+      // points to the original source file and an underlying/actual file path in a temporary
+      // directory. We're constrained by having to use only public APIs of the Java compiler
+      // and `toString()` seems to be the only way to access the user-friendly path.
+      if (toString.startsWith("SimpleFileObject[") && toString.endsWith("]")) {
+        return Paths.get(toString.substring("SimpleFileObject[".length(), toString.length() - 1));
+      } else {
+        throw new IllegalArgumentException("unsupported source file: " + toString);
+      }
     } else {
       return Paths.get(uri);
     }
   }
 
+  // Infers the `-sourceroot:` flag from the provided file
+  private void inferBazelSourceroot(JavaFileObject file) {
+    if (options.uriScheme != UriScheme.BAZEL || options.sourceroot != null) {
+      return;
+    }
+    Path absolutePath = absolutePathFromUri(options, file);
+    Path uriPath = Paths.get(file.toUri());
+    // absolutePath is the "human-readable" original path, for example
+    // /home/repo/com/example/Hello.java
+    // uriPath is the sandbox/temporary file path, for example
+    // /private/var/tmp/com/example/Hello.java
+    //
+    // We infer sourceroot by iterating the names of both files in reverse order and stop at the
+    // first entry where  the two paths are different. For the  example above, we compare
+    // "Hello.java", then "example",  then "com",  and when we reach  "repo" !+ "tmp" then we guess
+    // that "/home/repo" is the sourceroot. This  logic is brittle  and it would be nice to use more
+    // dedicated APIs, but Bazel actively makes an effort to  sandbox compilation  and hide access
+    // to the original workspace, which is why we resort to  solutions like this.
+    int relativePathDepth = 0;
+    int uriPathDepth = uriPath.getNameCount();
+    int absolutePathDepth = absolutePath.getNameCount();
+    while (relativePathDepth < uriPathDepth && relativePathDepth < absolutePathDepth) {
+      String uriName = uriPath.getName(uriPathDepth - relativePathDepth - 1).toString();
+      String pathName = absolutePath.getName(absolutePathDepth - relativePathDepth - 1).toString();
+      if (!uriName.equals(pathName)) {
+        break;
+      }
+      relativePathDepth++;
+    }
```

Comment:
suggestion(non-blocking): Perhaps add some unit tests for this function or specifically this logic? It seems a bit tricky.

Reply:
Good suggestion. I opened a reminder issue and a `// FIXME https://github.com/sourcegraph/lsif-java/issues/444` comment. Don't want to delay this PR further.